### PR TITLE
Use an actual DNS resolver

### PR DESCRIPTION
**Pull request by stayallive**
Without this the statistics requests will never reach the server if it's not available on 127.0.0.1 directly (think load balanced or distributed services). 

Let the system supply a DNS resolver or use Cloudflare's 1.1.1.1 as an default.

### CODE REVIEW

**Comment by mpociot** [comment]
The Problem is that the systems dns resolver will never resolve domains while developing locally. For example when using Laravel Valet. 
Maybe we could make this dependent on wether or not the app environment is set to production. 

What do you think @freekmurze ?

**Comment by stayallive** [comment]
@mpociot I noticed that too 👍 I made an config option for it 😄

**Comment by freekmurze** [comment]
Wouldn't that DNS lookup block our server?

**Comment by stayallive** [comment]
It shouldn't since it's using the [Reacts Async DNS lookup](https://github.com/reactphp/dns) and also uses the in-memory cache for lookups.

But otherwise some other way needs to be devised to get the statistics across since it just silently fails on our production cluster without it 😄

File: config/websockets.php [hunk 1]
```diff
@@ -65,6 +65,12 @@
          * the number of days specified here will be deleted.
          */
         'delete_statistics_older_than_days' => 60,
+
+        /*
+         * Use an DNS resolver to make the requests to the statistics logger
+         * default is to resolve everything to 127.0.0.1.
+         */
+        'dns_lookup' => false,
```
Comment (freekmurze) [inline config/websockets.php]: Rename this to `perform_dns_lookup`

File: src/Console/StartWebSocketServer.php [hunk 2]
```diff
@@ -46,8 +48,21 @@ public function handle()
...
+        if (config('websockets.statistics.dns_lookup')) {
+            $dnsConfig = DnsConfig::loadSystemConfigBlocking();
+
+            $dns = (new DnsFactory)->createCached(
+                $dnsConfig->nameservers
+                    ? reset($dnsConfig->nameservers)
+                    : '1.1.1.1',
+                $this->loop
+            );
+        } else {
```
Comment (freekmurze) [inline src/Console/StartWebSocketServer.php]: Avoid this else statement by refactoring to a dedicated method (that uses early returns) that 'll return a `DnsResolver`

**Comment by freekmurze** [comment]
I'm ok with adding this to the package. I don't know how, but it would be nice if we could have an automated test for this. 

Could you also PR that docs (update the config file there)

**Comment by stayallive** [comment]
Hmm we could test if it resolves not to 127.0.0.1 with that config option set possibly.

File: src/Console/StartWebSocketServer.php [hunk 3]
```diff
@@ -123,4 +125,20 @@ protected function startWebSocketServer()
             ->createServer()
             ->run();
     }
+
+    protected function getDnsResolver()
```
Comment (freekmurze) [inline src/Console/StartWebSocketServer.php]: Add a return type hint here

**Comment by WyriHaximus** [comment]
A small note FYI, we actually support the `hosts` file of *nix and Windows systems through the [`HostsFileExecutor`](https://reactphp.org/dns/#hostsfileexecutor) which is used by default on both the [non-cached and the cached resolvers](https://github.com/reactphp/dns/blob/75ae3286c03eb022a6fbf909e6148d5022588ba4/src/Resolver/Factory.php#L27-L68). And as long as you try to resolve `localhost` it will always return `127.0.0.1`.

**Comment by stayallive** [comment]
@WyriHaximus interesting... so as long as the entry is in the hosts file it will never even hit the network? Very nice!

This will still not allow us to enable it by default because Valet does not write to the hosts file but good to know that adding entries to the hosts file will make sure no network requests are made for DNS resolving when this _is_ enabled.

**Comment by mpociot** [comment]
@WyriHaximus Valet places a custom configuration, depending on the TLD that you choose inside the `etc/resolver/` folder. And this DNS will then point to 127.0.0.1. 
Too bad that it's not possible to add multiple DNS servers yet.

@stayallive thanks for the PR!

**Pull request merged**